Flaky-test: PulsarFunctionsTest.testExclamationFunction

org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest is flaky. The testExclamationFunction test method (used by multiple tests) fails sporadically.

Some old examples:
```
java.lang.AssertionError: expected [2] but found [1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.getFunctionStatus(PulsarFunctionsTest.java:2008)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testExclamationFunction(PulsarFunctionsTest.java:1649)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testJavaExclamationFunction(PulsarFunctionsTest.java:1564)
```

[example failure 2021-02-01T12:01:59.5076056Z](https://github.com/apache/pulsar/runs/1805264766?check_suite_focus=true#step:13:14331)
[example failure 2021-01-31T21:26:48.5466323Z](https://github.com/apache/pulsar/runs/1802273091?check_suite_focus=true#step:13:67327)
[example failure 2021-01-31T20:30:04.5865440Z](https://github.com/apache/pulsar/runs/1802273028?check_suite_focus=true#step:13:8533)
[example failure 2021-01-30T19:25:17.6626034Z](https://github.com/apache/pulsar/runs/1798266770?check_suite_focus=true#step:13:45042)
[example failure 2021-01-30T19:10:09.3467886Z](https://github.com/apache/pulsar/runs/1798266817?check_suite_focus=true#step:13:17160)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [2] but found [1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.getFunctionStatus(PulsarFunctionsTest.java:2008)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testExclamationFunction(PulsarFunctionsTest.java:1649)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testJavaExclamationFunction(PulsarFunctionsTest.java:1564)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

[example failure 2021-02-01T07:37:05.9664975Z](https://github.com/apache/pulsar/runs/1804234627?check_suite_focus=true#step:13:41386)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [2] but found [1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.getFunctionStatus(PulsarFunctionsTest.java:2008)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testExclamationFunction(PulsarFunctionsTest.java:1649)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testJavaExclamationFunction(PulsarFunctionsTest.java:1564)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

07:37:05.935 [docker-java-stream-455335145:org.apache.pulsar.tests.integration.utils.DockerUtils$2@221] INFO  org.apache.pulsar.tests.integration.utils.DockerUtils - DOCKER.exec(PulsarFunctionsProcessTest-ddftp-pulsar-broker-0:tail -f /var/log/pulsar/broker.log): STDOUT: 07:37:05.735 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.ServerCnx - [PersistentTopic{topic=persistent://public/default/test-tumbling-count-window-PROCESS-input-vzgsrlqj}][PulsarFunctionsProcessTest-ddftp-0-28] Closing producer on cnx /172.20.0.8:43662. producerId=0
07:37:05.735 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.ServerCnx - [PersistentTopic{topic=persistent://public/default/test-tumbling-count-window-PROCESS-input-vzgsrlqj}][PulsarFunctionsProcessTest-ddftp-0-28] Closed producer on cnx /172.20.0.8:43662. producerId=0
07:37:05.738 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.ServerCnx - [/172.20.0.8:43662] Closing consumer: consumerId=0
07:37:05.738 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.AbstractDispatcherSingleActiveConsumer - Removing consumer Consumer{subscription=PersistentSubscription{topic=persistent://public/default/test-tumbling-count-window-PROCESS-output-hznrrdkg, name=reader-0ed355cd16}, consumerId=0, consumerName=e6cf3, address=/172.20.0
07:37:05.935 [docker-java-stream-455335145:org.apache.pulsar.tests.integration.utils.DockerUtils$2@221] INFO  org.apache.pulsar.tests.integration.utils.DockerUtils - DOCKER.exec(PulsarFunctionsProcessTest-ddftp-pulsar-broker-0:tail -f /var/log/pulsar/broker.log): STDOUT: .8:43662}
07:37:05.738 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.persistent.PersistentSubscription - [persistent://public/default/test-tumbling-count-window-PROCESS-output-hznrrdkg][reader-0ed355cd16] Successfully closed subscription [NonDurableCursorImpl{ledger=public/default/persistent/test-tumbling-count-window-PROCESS-output-hznrrdkg, ackPos=146:2, readPos=146:3}]
07:37:05.738 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.persistent.PersistentSubscription - [persistent://public/default/test-tumbling-count-window-PROCESS-output-hznrrdkg][reader-0ed355cd16] Successfully closed dispatcher for reader
07:37:05.738 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.ServerCnx - [/172.20.0.8:43662] Closed consumer, consumerId=0
07:37:05.747 [pulsar-io-25-2] INFO  org.apache.pulsar.broker.service.ServerCnx - Closed connection from /172.20.0.8:43656
07:37:05.749 [pulsar-io-25-3] INFO  org.apache.pulsar.broker.service.ServerCnx - Closed connection from /172.20.0.8:43662
</pre></code>
</details>

```
java.lang.AssertionError: expected [2] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.getFunctionStatus(PulsarFunctionsTest.java:1991)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testExclamationFunction(PulsarFunctionsTest.java:1647)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testJavaExclamationFunction(PulsarFunctionsTest.java:1562)
```

[example failure 2021-01-28T21:11:11.2781744Z](https://github.com/apache/pulsar/runs/1786879847?check_suite_focus=true#step:12:17989)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [2] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.getFunctionStatus(PulsarFunctionsTest.java:1991)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testExclamationFunction(PulsarFunctionsTest.java:1647)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testJavaExclamationFunction(PulsarFunctionsTest.java:1562)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

```
java.lang.AssertionError: expected [2] but found [1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.getFunctionStatus(PulsarFunctionsTest.java:1991)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testExclamationFunction(PulsarFunctionsTest.java:1647)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testJavaExclamationFunction(PulsarFunctionsTest.java:1562)
```

[example failure 2021-01-28T16:08:05.8990897Z](https://github.com/apache/pulsar/runs/1784405246?check_suite_focus=true#step:12:44025)
[example failure 2021-01-28T15:52:51.9589532Z](https://github.com/apache/pulsar/runs/1784405246?check_suite_focus=true#step:12:29580)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [2] but found [1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.getFunctionStatus(PulsarFunctionsTest.java:1991)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testExclamationFunction(PulsarFunctionsTest.java:1647)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testJavaExclamationFunction(PulsarFunctionsTest.java:1562)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
@@ -24,7 +24,6 @@ import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertTrue;
 import static org.testng.Assert.fail;
 import com.google.gson.Gson;
-
 import java.time.Duration;
 import java.util.Collections;
 import java.util.HashSet;
@@ -35,7 +34,6 @@ import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
-
 import lombok.Cleanup;
 import lombok.extern.slf4j.Slf4j;
 import net.jodah.failsafe.Failsafe;
@@ -93,6 +91,7 @@ import org.apache.pulsar.tests.integration.io.SourceTester;
 import org.apache.pulsar.tests.integration.topologies.FunctionRuntimeType;
 import org.apache.pulsar.tests.integration.topologies.PulsarCluster;
 import org.assertj.core.api.Assertions;
+import org.awaitility.Awaitility;
 import org.testcontainers.containers.GenericContainer;
 import org.testcontainers.shaded.com.google.common.collect.Sets;
 import org.testng.annotations.Test;
@@ -1647,8 +1646,13 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
         // update parallelism
         updateFunctionParallelism(functionName, 2);
 
-        //get function status
-        getFunctionStatus(functionName, 0, true, 2);
+        Awaitility.await()
+                .pollInterval(Duration.ofMillis(500L))
+                .ignoreExceptions()
+                .untilAsserted(() ->
+                //get function status
+                getFunctionStatus(functionName, 0, true, 2)
+        );
 
         // delete function
         deleteFunction(functionName);

